doc/src/guides/solving/solve-equation-algebraically.md

```diff
@@ -0,0 +1,394 @@
+>>> solutions = solve(parsed, x, dict=True)
+>>> print(solutions)
+[{x: -sqrt(y)}, {x: sqrt(y)}]
+>>> for solution in solutions:
+...     for key, val in solution.items():
+...         print(val)
+-sqrt(y)
+sqrt(y)
+```
+
+### Restricting the domain of solutions using {func}`~.solve`
+
+By default, SymPy will return solutions in the complex domain, which also includes 
+purely real and imaginary values. Here, the first two solutions are real, 
+and the last two are imaginary:
+
+```py
+>>> from sympy import Symbol, solve
+>>> x = Symbol('x')
+>>> solution = solve(x ** 4 - 256, x)
+>>> print(solution)
+[-4, 4, -4*I, 4*I]
+```
+
+If you want to restrict returned solutions to real numbers, you can place an assumption on the 
+symbol to be solved for, $x$:
+
+```py
+>>> from sympy import Symbol, solve
+>>> x = Symbol('x', real=True)
+>>> solution = solve(x ** 4 - 256, x)
+>>> print(solution)
+[-4, 4]
+```
+
+### Substituting solutions from {func}`~.solve` into an expression
+
+You can substitute solutions from {func}`~.solve` into an expression.
+
+A common use case is finding the critical points and values for a function $f$. 
+At the critical points, the derivative equals zero (or is undefined). 
+You can then obtain the function values at those critical points
+by substituting the critical points back into the function using {meth}`~sympy.core.basic.Basic.subs`.
+
+```py
+>>> from sympy.abc import x
+>>> from sympy import solve, diff
+>>> f = x**3 + x**2 - x
+>>> derivative = diff(f, x)
+>>> critical_points = solve(derivative, x, dict=True)
+>>> print(critical_points)
+[{x: -1}, {x: 1/3}]
+>>> for critical_point in critical_points:
+...     # Extract the value from the dictionary where key is x
+...     critical_point_value = critical_point[x]
+...     f_at_critical_point = f.subs(x, critical_point_value)
+...     print(f"f({critical_point_value}) = {f_at_critical_point}")
+f(-1) = 1
+f(1/3) = -5/27
+```
+
+## Using {func}`~.solveset`
+
+- produces outputs in the format of 
+[SymPy mathematical Sets](https://docs.sympy.org/dev/modules/sets.html?highlight=sets#module-sympy.sets.sets) rather than 
+[Python sets](https://docs.python.org/3/library/stdtypes.html#set)
+- can return infinitely many solutions
+- the solution set can be more difficult to parse programmatically
+
+You can solve an equation using solve() in several ways.
+
+### Make your equation into an expression that equals zero
+
+Use the fact that any expression not in an `Eq` (equation) is automatically assumed to equal zero (0)
+ by the solving functions. You can rearrange the equation $x^2 = y$ to $x^2 - y = 0$, and 
+ {func}`~.solveset` that expression. This approach is convenient if you are interactively solving an 
+expression which already equals zero, or an equation that you do not mind rearranging to $expression = 0$.
+
+```py
+>>> from sympy import solveset
+>>> from sympy.abc import x, y
+>>> solution = solveset(x**2 - y, x)
+>>> print(solution)
+{-sqrt(y), sqrt(y)}
```

Comment:
Question number 1 to be answered here is how do I actually get the expressions out of this object?
```
In [11]: solution[0]
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-11-ed0f89115762> in <module>
----> 1 solution[0]

TypeError: 'FiniteSet' object is not subscriptable
```

Reply:
Good question. I learned that iterating through a finite set returns its expressions.

Reply:
But you don't always get a FiniteSet:
```python
In [1]: x, y = symbols('x, y')

In [2]: solveset(x**2 - y, x, Reals)
Out[2]: ℝ ∩ {-√y, √y}

In [3]: list(_)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-3-1a5b60e28d67> in <module>
----> 1 list(_)

~/current/sympy/sympy.git/sympy/sets/sets.py in __iter__(self)
   1554             if not candidates:
   1555                 raise TypeError("None of the constituent sets are iterable")
-> 1556             raise TypeError(
   1557                 "The computation had not completed because of the "
   1558                 "undecidable set membership is found in every candidates.")

TypeError: The computation had not completed because of the undecidable set membership is found in every candidates.
```
The question is what to do? Here if you declare `y` to be positive you'll get the FiniteSet:
```python
In [4]: y = symbols('y', positive=True)

In [5]: solveset(x**2 - y, x, Reals)
Out[5]: {-√y, √y}
```
It's not always possible in general to set appropriate assumptions though so the question is how do you detect if solveset returned a FiniteSet and what do you do about it when it doesn't? This is actually one of the main issues that users have with using `solveset` and also the main reason that many users prefer `solve` so I think this guide needs to address it.